types/src/bytesrepr.rs

```diff
@@ -670,6 +684,56 @@ where
     }
 }
 
+#[cfg(feature = "std")]
+impl<K, V> ToBytes for HashMap<K, V>
+where
+    K: ToBytes,
+    V: ToBytes,
+{
+    fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let mut buf = allocate_buffer(self)?;
+        self.write_bytes(&mut buf)?;
+        Ok(buf)
+    }
+
+    fn serialized_length(&self) -> usize {
+        U32_SERIALIZED_LENGTH
+            + self
+                .iter()
+                .map(|(key, value)| key.serialized_length() + value.serialized_length())
+                .sum::<usize>()
+    }
+
+    fn write_bytes(&self, writer: &mut Vec<u8>) -> Result<(), Error> {
+        let length_32: u32 = self.len().try_into().map_err(|_| Error::NotRepresentable)?;
+        writer.extend_from_slice(&length_32.to_le_bytes());
+        for (key, value) in self.iter() {
+            key.write_bytes(writer)?;
+            value.write_bytes(writer)?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K, V> FromBytes for HashMap<K, V>
+where
+    K: FromBytes + Hash + Ord,
+    V: FromBytes,
+{
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), Error> {
+        let (num_keys, mut stream) = u32::from_bytes(bytes)?;
+        let mut result = HashMap::new();
```

Comment:
We should use `with_capacity(usize::from(num_keys))` here, I think.

Reply:
Done.

`usize` has `TryFrom` instead of `From` for `u32`. Seems like 16-bit systems are still a thing.

Reply:
I'd be careful here with `with_capacity` @marc-casperlabs - `num_keys` is potentially untrusted data, and if somehow the host tries to deserialize `u32::MAX` here on the host, it will abort the process. In `ToBytes for Vec<T>,` we use `try_vec_with_capacity` to avoid this problem.

Reply:
> usize has TryFrom instead of From for u32. Seems like 16-bit systems are still a thing.

I specifically discounted them, I don't see us supporting those any time soon. But I guess it can't hurt _much_ to use `TryFrom` and we'll be happy to have done that once we port everything to the SNES ❤️ .

@mpapierski has a very good poin point that I missed though --- I forgot that there _is_ a limit since the incoming chunk of raw serialized data is likely size limited. So I guess no `with_capacity` (or a compromise using a small upper bound, but at that point, we might as well not bother).

tl;dr my bad, my suggestions weren't exactly stellar.

Reply:
Reverted the commit with the change.